src/reflect/scala/reflect/internal/Symbols.scala

```diff
@@ -840,7 +840,7 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
     final def isDelambdafyFunction = isSynthetic && (name containsName tpnme.DELAMBDAFY_LAMBDA_CLASS_NAME)
     final def isDelambdafyTarget  = isArtifact && isMethod && hasAttachment[DelambdafyTarget.type]
     final def isDefinedInPackage  = effectiveOwner.isPackageClass
-    final def needsFlatClasses    = phase.flatClasses && (rawowner ne NoSymbol) && !rawowner.isPackageClass
+    final def needsFlatClasses    = !isMethod && phase.flatClasses && (rawowner ne NoSymbol) && !rawowner.isPackageClass
```

Comment:
`isMethod` is actually a virtual call as things are currently factored. We could push this check to the end of this line (typically `phase.flatClasses` is mostly false).

As a followup task, we should review the way `isMethod` etc etc are factored with the help of micro-benchmarks and JITWatch analysis.

Reply:
I've pushed the condition reordering as an extra commit to this PR.